### PR TITLE
Feature/ch12312 - reorg attack protection in VRF jpv2

### PR DESCRIPTION
**Pull request by connorwstein**
The approach taken is the following:
1. Additionally subscribe to VRF fulfillment logs and maintain a map of requestID->fulfillment count
2. When receiving a VRF request log, add it to a queue with confirmedAt = minConfs\*(2^(fulfillment count)). So a new request will have confirmedAt = minConfs, an already fulfilled request will have confirmedAt=minConfs\*2, one fulfilled twice will have confirmedAt=minConfs\*4 and so on. This way we only double if the output has actually been seen before reorging the request. Note however that it does not take into account whether the request has actually been moved to a different block (thus changing the output), so its possible that on a back-and-forth longest chain tip scenario we would unnecessarily double. That seems acceptable as we cap the doubling at 200 confs. 
3. Additionally listen to new heads. Process the queue of the requests as they become confirmed. Also prune any fulfillment map entries older than 10000 blocks to avoid that map growing forever.

### CODE REVIEW

File: core/services/vrf/delegate.go [hunk 1]
```diff
@@ -357,13 +537,14 @@ func GetVRFInputs(jb job.Job, request *solidity_vrf_coordinator_interface.VRFCoo
 func (lsn *listener) Close() error {
 	return lsn.StopOnce("VRFListener", func() error {
 		close(lsn.chStop)
-		<-lsn.waitOnStop
+		<-lsn.waitOnStop // Log listener
+		<-lsn.waitOnStop // Head listener
```
Comment (PiotrTrzpil) [inline core/services/vrf/delegate.go:541]: could use a waitgroup? seems to be a pattern in other services

File: core/services/vrf/delegate.go [hunk 2]
```diff
@@ -218,72 +375,96 @@ func (lsn *listener) run(unsubscribeLogs func(), minConfs uint32) {
...
+				lsn.reqAdded()
+				lsn.reqsMu.Unlock()
 			}
 		}
 	}
 }
 
+func (lsn *listener) getConfirmedAt(req *solidity_vrf_coordinator_interface.VRFCoordinatorRandomnessRequest, minConfs uint32) uint64 {
+	newConfs := uint64(minConfs) * (1 << lsn.respCount[req.RequestID])
+	if newConfs > 200 {
```
Comment (PiotrTrzpil) [inline core/services/vrf/delegate.go:394]: hm i think this was already here, but might be good to add a comment or a named constant for why 200

File: core/services/vrf/delegate_test.go [hunk 3]
```diff
@@ -88,91 +108,226 @@ func generateCallbackReturnValues(t *testing.T) []byte {
...
+				BlockNumber: 10,
+				BlockHash:   bh,
+			},
 		},
 	}
 
 	consumed := make(chan struct{})
-	for i, l := range logs {
+	for i, tc := range tt {
+		tc := tc
```
Comment (PiotrTrzpil) [inline core/services/vrf/delegate_test.go:330]: might be more readable to have some longer names than tt and tc

File: core/internal/testutils/pgtest/pgtest.go [hunk 4]
```diff
@@ -27,6 +27,7 @@ func NewGormDB(t *testing.T) *gorm.DB {
 	sqlDB := NewSqlDB(t)
 	gormDB, err := gorm.Open(postgres.New(postgres.Config{
 		Conn: sqlDB,
+		//DSN: uuid.NewV4().String(),
```
Comment (PiotrTrzpil) [inline core/internal/testutils/pgtest/pgtest.go:30]: to remove?

File: core/services/vrf/delegate.go [hunk 5]
```diff
@@ -218,72 +375,96 @@ func (lsn *listener) run(unsubscribeLogs func(), minConfs uint32) {
...
+			if err != nil {
+				return errors.Wrap(err, "VRFListener: failed to insert finished run")
+			}
+		}
+		// Always mark consumed regardless of whether the proof failed or not.
+		err = lsn.logBroadcaster.MarkConsumed(tx, lb)
+		if err != nil {
+			return err
+		}
+		return nil
```
Comment (PiotrTrzpil) [inline core/services/vrf/delegate.go:461]: minor, but you could just do return `lsn.logBroadcaster.MarkConsumed(tx, lb)`

File: core/services/vrf/delegate.go [hunk 6]
```diff
@@ -122,24 +137,59 @@ var (
...
+}
+
+func (lsn *listener) Connect(head *models.Head) error {
+	lsn.latestHead = uint64(head.Number)
+	return nil
+}
+
+// Note that we have 2 seconds to do this processing
+func (lsn *listener) OnNewLongestChain(ctx context.Context, head models.Head) {
+	lsn.latestHead = uint64(head.Number)
```
Comment (PiotrTrzpil) [inline core/services/vrf/delegate.go:188]: hm seems like `lsn.latestHead` is accessed in multiple goroutines - might need to be an atomic?

File: core/services/vrf/delegate.go (hunk 1 above)
Comment (connorwstein) [reply core/services/vrf/delegate.go:577]: yeah I was going to but its actually nice to have the channel for testability https://github.com/smartcontractkit/chainlink/blob/4803aba4099ab2f7b2e800891fc1d343a4ed0285/core/services/vrf/delegate_test.go#L163

File: core/services/vrf/delegate_test.go (hunk 3 above)
Comment (connorwstein) [reply core/services/vrf/delegate_test.go:400]: meh in the go docs they use tt := range tests, I like tc := range tt (tc = test case)

Comment (PiotrTrzpil) [reply core/services/vrf/delegate_test.go:397]: ah ok yeah makes sense

**Review by PiotrTrzpil — APPROVED** [review]

**Comment by PiotrTrzpil** [comment]
looks great!

**Pull request merged**